src/core/utils/flash.cpp

```diff
@@ -167,53 +394,45 @@ otError Flash::Get(uint16_t aKey, int aIndex, uint8_t *aValue, uint16_t *aValueL
 
 otError Flash::Set(uint16_t aKey, const uint8_t *aValue, uint16_t aValueLength)
 {
-    return Add(aKey, true, aValue, aValueLength);
+    Delete(aKey, -1);
```

Comment:
Excellent point. We could:
1. Delete after adding (we have to think what to do in some situations, I think it's not as easy as it seems)
2. Restore the `first` flag.
3. Think of another alternative!

Reply:
With "delete after adding", we would also need to handle the case that there would be multiple entries after we write the new entry but before we delete the old entry.

I do not have a better proposal than the `first` flag right now, but I am open to other ideas if you have them.

Reply:
Yes, that's the case I was thinking of. Every solution I can imagine seems to use something similar to the `first` flag.

Reply:
Implemented an idea similar to the `First` flag, but instead of meaning "first", it means "invalidate previous" records. 

This is a sublte difference, but avoids the need to mark "first" the record `index == 1` when deleting record `index == 0`. This retains compatibility with the old format's `First` flag. Also, the `InvalidatePrevious` flag implicitly resets the iterator's index to 0. Please see c5f19dac689266ad9f04e2df69cd4cbd3593bebf.